src/CoreWCF.Primitives/src/CoreWCF/Security/LdapAdapter.cs

```diff
@@ -0,0 +1,119 @@
+
+using System;
+using System.Collections.Generic;
+using System.DirectoryServices.Protocols;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using CoreWCF.IdentityModel.Claims;
+using Microsoft.Extensions.Caching.Memory;
+
+namespace CoreWCF.Security
+{
+    internal static class LdapAdapter
+    {
+        public static async Task<List<Claim>> RetrieveClaimsAsync(LdapSettings settings, string originalUserName)
+        {
+            var upnIndex = originalUserName.IndexOf('@');
+            var userAccountName = "";
+            if (upnIndex == -1)
+            {
+                int domainIndex = originalUserName.IndexOf("\\");
+                if (domainIndex == -1)
+                    return null;
+                userAccountName = originalUserName.Substring(domainIndex + 1);
+            }
+            else
+            {
+                userAccountName = originalUserName.Substring(0, upnIndex);
+            }
+
+            List<Claim> roleClaims = new List<IdentityModel.Claims.Claim>();
+
+            if (settings.ClaimsCache.TryGetValue<IEnumerable<string>>(originalUserName, out var cachedClaims))
+            {
+                foreach (var claim in cachedClaims)
+                {
+                    roleClaims.Add(new Claim(ClaimTypes.Role, claim, Rights.Identity));
+                }
+
+                return roleClaims;
+            }
+
+            var distinguishedName = settings.Domain.Split('.').Select(name => $"dc={name}").Aggregate((a, b) => $"{a},{b}");
+            var retrievedClaims = new List<string>();
+            if (!string.IsNullOrEmpty(settings.OrgUnit))
+            {
+                distinguishedName = "OU=" + settings.OrgUnit + "," + distinguishedName;
+            }
+
+            var genericFilter = $"(&(objectClass=user)(sAMAccountName={userAccountName}))"; // This is using ldap search query language, it is looking on the server for someUser
+            var upnFilter = $"(&(objectClass=user)(userPrincipalName={originalUserName}))";
+            var genericSearchRequest = new SearchRequest(distinguishedName, genericFilter, SearchScope.Subtree, null);
+            var upnSearchRequest = new SearchRequest(distinguishedName, upnFilter, SearchScope.Subtree, null);
+            SearchResponse searchResponse = null;
+            try
+            {
+                if (upnIndex > 0)
+                {
+                    searchResponse = (SearchResponse)await Task<DirectoryResponse>.Factory.FromAsync(
+                    settings.LdapConnection.BeginSendRequest, settings.LdapConnection.EndSendRequest,
+                    upnSearchRequest, PartialResultProcessing.NoPartialResultSupport, null);
+                    if(searchResponse !=null && searchResponse.Entries != null && searchResponse.Entries.Count > 1)
+                    {
+                        throw new Exception(SR.DuplicateUPN); //resource
+                    }
```

Comment:
We need to find a better exception than just Exception. Maybe SecurityNegotiationException? Maybe a new exception type needs to be created?

Reply:
SecurityNegotiationException